flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinatorTest.java

```diff
@@ -624,6 +625,36 @@ public void testTriggerAndDeclineCheckpointThenFailureManagerThrowsException()
         }
     }
 
+    @Test
+    public void testIOExceptionCheckpointExceedsTolerableFailureNumber() throws Exception {
```

Comment:
This test checks that the task is failed when it receives the new reason type but I don't see tests that check that your new code inside of `onTriggerFailure` and `onTriggerFailure` work. 
As I can see, at least `testPeriodicSchedulingWithInactiveTasks`, `testTriggerCheckpointAfterCancel` and `testCheckpointAbortsIfTriggerTasksAreFinished`(and I sure many more) trigger somehow the exception. Perhaps, you can take a look how these tests works and you can write something similar.

Reply:
Sure, I have already add some unit tests that check getCheckpointException function.